src/dag.rs

```diff
@@ -30,43 +174,414 @@ impl<Types: RepoTypes> IpldDag<Types> {
+    async fn resolve_hamt(
+        &self,
+        mut lookup: ShardedLookup<'_>,
+        cache: &mut Option<Cache>,
+    ) -> Result<Cid, Error> {
+        use MaybeResolved::*;
+
+        loop {
+            let (next, _) = lookup.pending_links();
+
+            let block = self.ipfs.repo.get_block(next).await?;
+
+            match lookup.continue_walk(block.data(), cache)? {
+                NeedToLoadMore(next) => lookup = next,
+                Found(cid) => return Ok(cid),
+                NotFound => return Err(anyhow::anyhow!("key not found: ???")),
+            }
         }
-        Ok(ipld)
-    }
-}
-
-fn try_resolve(ipld: Ipld, segment: &str) -> Result<Ipld, Error> {
-    match ipld {
-        Ipld::Map(mut map) => map
-            .remove(segment)
-            .ok_or_else(|| anyhow::anyhow!("no such segment: {:?}", segment)),
-        Ipld::List(mut vec) => match segment.parse::<usize>() {
-            Ok(index) if index < vec.len() => Ok(vec.swap_remove(index)),
-            Ok(_) => Err(anyhow::anyhow!(
-                "index out of range 0..{}: {:?}",
-                vec.len(),
-                segment
-            )),
-            Err(_) => Err(anyhow::anyhow!("invalid list index: {:?}", segment)),
-        },
-        link @ Ipld::Link(_) if segment == "." => Ok(link),
-        other => Err(anyhow::anyhow!(
-            "cannot resolve {:?} through: {:?}",
+    }
+}
+
+/// `IpfsPath`'s cid based variant can be resolved to the block, projections represented by this
+/// type.
+///
+/// Values can be converted to Ipld using `Ipld::try_from`.
+#[derive(Debug, PartialEq)]
+pub enum ResolvedNode {
+    /// Block which was loaded at the end of the path.
+    Block(Block),
+    /// Path ended in `Data` at a dag-pb node. This is usually not interesting and should be
+    /// treated as "Not found" error since dag-pb node did not have a *link* called `Data`. Variant
+    /// exists as there are interface-ipfs-http tests which require this behaviour.
+    DagPbData(Cid, NodeData<Box<[u8]>>),
+    /// Path ended on an !dag-pb document which was projected.
+    Projection(Cid, Ipld),
+    /// Local resolving ended in at link
+    Link(Cid, Cid),
+}
+
+impl ResolvedNode {
+    /// Returns the cid of the **source** document. Source or destination matters only the case of link
+    /// to another document.
+    pub fn source(&self) -> &Cid {
+        match self {
+            ResolvedNode::Block(Block { cid, .. })
+            | ResolvedNode::DagPbData(cid, ..)
+            | ResolvedNode::Projection(cid, ..)
+            | ResolvedNode::Link(cid, ..) => cid,
+        }
+    }
+
+    /// Returns the cid of the linked document when resolved to a link, otherwise `None`.
+    pub fn destination(&self) -> Option<&Cid> {
+        match self {
+            ResolvedNode::Link(_, cid) => Some(cid),
+            _ => None,
+        }
+    }
+
+    /// Returns the destination or the source link.
```

Comment:
nit: the docs are a bit quirky here,  as `fn destination` states `Returns the cid` and `fn cid` (which is right next to it) states `Returns the destination`; as long as the meaning is solid, though, it's not a big deal

Reply:
These are my own inventions however. `source` is the source for the `ResolvedNode` variants, including the link, however the link has a second `Cid` as `destination`. I can't remember from where I refactored these, possibly from the http. Yeah it was http, no longer useful since I pushed the path resolving downwards, and `source()` would be enough.

```diff
diff --git a/http/src/v0/dag.rs b/http/src/v0/dag.rs
index 548be12..2db6bd2 100644
--- a/http/src/v0/dag.rs
+++ b/http/src/v0/dag.rs
@@ -140,7 +140,7 @@ async fn inner_resolve<T: IpfsTypes>(
         .map_err(StringError::from)?
         .map_err(StringError::from)?;

-    let current = resolved.cid();
+    let current = resolved.source();

     Ok(reply::json(&json!({
         "Cid": { "/": current.to_string() },
diff --git a/src/dag.rs b/src/dag.rs
index 185ff01..6b8d796 100644
--- a/src/dag.rs
+++ b/src/dag.rs
@@ -50,9 +50,9 @@ pub enum ResolveError {

 #[derive(Debug, Error)]
 pub enum UnexpectedResolved {
-    #[error("path resolved to unexpected type of document: {:?} or {}", .0, .1.cid())]
+    #[error("path resolved to unexpected type of document: {:?} or {}", .0, .1.source())]
     UnexpectedCodec(cid::Codec, ResolvedNode),
-    #[error("path did not resolve to a block on {}", .0.cid())]
+    #[error("path did not resolve to a block on {}", .0.source())]
     NonBlock(ResolvedNode),
 }

@@ -344,23 +344,10 @@ impl ResolvedNode {
         }
     }

-    /// Returns the cid of the linked document when resolved to a link, otherwise `None`.
-    pub fn destination(&self) -> Option<&Cid> {
-        match self {
-            ResolvedNode::Link(_, cid) => Some(cid),
-            _ => None,
-        }
-    }
-
-    /// Returns the destination or the source link.
-    pub fn cid(&self) -> &Cid {
-        self.destination().unwrap_or_else(|| self.source())
-    }
-
     /// Unwraps the dagpb block variant and turns others into UnexpectedResolved.
     /// This is useful wherever unixfs operations are continued after resolving an IpfsPath.
     pub fn into_unixfs_block(self) -> Result<Block, UnexpectedResolved> {
-        if self.cid().codec() != cid::Codec::DagProtobuf {
+        if self.source().codec() != cid::Codec::DagProtobuf {
             Err(UnexpectedResolved::UnexpectedCodec(
                 cid::Codec::DagProtobuf,
                 self,
```

Would it be more clear to have only the source remain?

Reply:
I am always for having less API surface but does this clarify the root cause?